### ISSUE TITLE
Tool availability detection misreports installed tools as missing

### ISSUE THREAD

**Issue by MarcusJellinghaus** [MarcusJellinghaus/mcp-code-checker#220]
## Symptom

`run_pytest_check` always returns:

```
pytest is not available in the configured Python environment
(C:\...\mcp_coder_1045\.venv\Scripts\python.exe). Ensure --python-executable
and --venv-path point to the environment where pytest is installed.
Restart the server after installing.
```

pytest **is** installed in that exact interpreter (pytest 8.4.2), and `python -m pytest --version` succeeds with exit code 0. The message is wrong, and it points at a configuration fix that cannot help.

## Root cause

Tool availability is detected two different ways, and both key on the wrong thing.

**Path 1 — lazy subprocess probe** (`server.py:214-239`), used for pytest, pylint, mypy, black, isort:

```python
result = execute_command(
    [self._resolved_python, "-m", tool_name, "--version"],
    timeout_seconds=10,
)
available = result.return_code == 0 and not result.execution_error
```

On this project the probe simply takes longer than 10s:

| Probe | Time |
|---|---|
| `python -m pytest --version` | 15.2s / 14.6s / 14.9s (3 consecutive runs) |
| same, with `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1` | 5.2s |

The ~10s difference is pytest plugin autoload. The project has a substantial plugin set (xdist among them), and each is imported during startup even for `--version`.

**Path 2 — eager file existence** (`server.py:115-212`), used for lint-imports, vulture, ruff, bandit, tach: checks `os.path.exists(venv_path/Scripts/<tool>.exe)`. This has a separate, unreported bug — it is gated entirely on `self.venv_path`, so when that is `None` all five tools are recorded **unavailable**. `--venv-path` is optional; `--python-executable` alone and bare `sys.executable` are both supported configurations.

Same defect in both paths: detection is keyed on something other than the interpreter that will actually be used.

**Environment note:** this is a not-especially-fast Windows laptop. That is precisely the point — the failure is a function of machine speed and plugin count, so it will hit slower hardware, large plugin sets, cold filesystem caches and busy CI runners, while remaining invisible on a fast developer machine. 10s is a reasonable-looking number with very little margin: even with autoload disabled, bare interpreter + pytest startup is already 5.2s here.

## Why this is worse than a slow probe

1. **It fails closed and silently.** The tool is not slow — it becomes permanently unavailable.
2. **The negative result is cached** (`self._tool_availability[tool_name] = available`), so a single slow probe disables the tool for the whole server lifetime. No retry, no recovery without a restart.
3. **The message misdiagnoses.** "Ensure `--python-executable` and `--venv-path` point to the environment where pytest is installed" sends you to check configuration that is already correct.
4. **There is no workaround from the client side.** `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1` would make the probe fit, but it also disables xdist, so `-n auto` — which this project's own guidance mandates — stops working.

## `--venv-path` is redundant with `--python-executable`

`self.python_executable` is read in exactly one place: `_resolve_python_executable`. `self.venv_path` is read in three — that same resolution, `_check_tool_availability`, and the pytest call site that feeds the runner's `PATH` prepend (`checker_tools/pytest_tool.py:106`). All three immediately reduce it to the venv's script directory, which is the directory `_resolved_python` already lives in.

Both real `.mcp.json` files pass both flags, pointing at the same venv:

```json
"--python-executable", "${VIRTUAL_ENV}\\Scripts\\python.exe",
"--venv-path",         "${VIRTUAL_ENV}",
```

They can disagree. mcp-config makes two *independent* detections (`servers.py:165-175`): `auto_detect_python_executable` calls `detect_python_environment`, which falls back to `sys.executable` when the venv python fails `validate_python_executable`, while `auto_detect_venv_path` calls `find_virtual_environments` and returns `venvs[0]` unconditionally. The two diverge when the first-found venv has a missing or non-validating python — conda is a concrete case, since `get_venv_python` looks for `Scripts/python.exe` while conda puts the interpreter at the env root. In practice both flags point at the same venv, so `_resolve_python_executable`'s "venv wins" precedence has not mattered — but it is not structurally guaranteed. Decision 11 preserves today's precedence regardless, so the impact is low.

## Approach

Replace the single `tool_name` string with a per-tool descriptor (`key` / `script` / `module`), and collapse `_check_tool_availability` — currently the same 12-line block copy-pasted five times — into a table plus a loop.

| key | script | module |
|---|---|---|
| pytest | `pytest` | `pytest` |
| pylint | `pylint` | `pylint` |
| mypy | `mypy` | `mypy` |
| black | `black` | `black` |
| isort | `isort` | `isort` |
| lint-imports | `lint-imports` | — |
| vulture | `vulture` | — |
| ruff | `ruff` | — |
| bandit | `bandit` | — |
| tach | `tach` | — |

Detection derives the script directory from `_resolved_python`, cheapest check first:

1. **Console script exists** next to `_resolved_python` → available. For the five script tools, store its path in `self._tool_binaries[key]`; pytest/pylint/mypy/black/isort are always invoked as `python -m <module>`, so no entry is stored for them. No subprocess. Instant in any venv, on either platform.
2. **Otherwise, if the descriptor has a `module`**, probe `python -m <module> --version`, timeout `30`, with `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1` set on the probe subprocess only. No `module` → unavailable.
3. **Timeout → treat as available**, cache that, log a WARNING naming the timeout and the interpreter.

Predicate, in this order:

```
timed_out                        → True   (WARNING, cached)
rc == 0 and not execution_error  → True
otherwise                        → False  (cached)
```

`--venv-path` is soft-deprecated: it still resolves the interpreter with today's precedence, is no longer used for detection or for pytest's `PATH` prepend, logs a deprecation warning, and is hidden from `--help`.

## Constraints & Rationale

- **The five script tools cannot fail open.** `_check_tool_availability` does not only decide availability — it produces the path the tool is *run* with (`binary = server._ruff_binary` / `assert binary is not None` / `run_ruff_check_impl(ruff_binary=binary, ...)`, and the same shape in the other four). A probe returning `True` with no script on disk would leave that path `None`, trip the assert, and surface `Unexpected error running ruff check: AssertionError:` — worse than the message being fixed. Hence `module=None` for all five: file existence is the whole check, and "available" always implies "we hold a concrete path".

- **One `tool_name` string can no longer serve.** It is currently overloaded as dict key, module name for `python -m`, *and* console-script filename. `lint-imports` is where they diverge: script `lint-imports`, and `python -m lint-imports` is impossible because module names cannot contain hyphens. Today this is masked by an accident — `lint-imports` is always in `_tool_availability` before `_is_tool_available` runs, so the probe branch is unreachable for it. Do not inherit that accident. `.replace("-", "_")` is not a workaround: hyphen/underscore equivalence is PEP 503 and applies to distribution names only. The split is still required even though lint-imports ends up with `module=None` — the table is what makes that deliberate rather than accidental.

- **File existence proves presence; absence proves nothing — for the probe group only.** In a venv the interpreter and console scripts share a directory, on both platforms, so the fast path answers instantly. For a non-venv interpreter they do not (Windows `C:\Python311\python.exe` vs `C:\Python311\Scripts\pytest.exe`; POSIX `/usr/bin/python3` vs `/usr/local/bin/pytest`), hence the `python -m` fallback. The five script tools have no such fallback by construction, so a non-venv interpreter still reports them unavailable — unchanged from today, and out of scope here.

- **Fail open on timeouts only, and cache it.** A timeout proves the probe useless for that tool on that machine, not that the tool is absent. Returning "unavailable but uncached" (the obvious reading of "don't cache a timeout") still returns the wrong answer this call *and* re-pays the timeout on every later call — at 30s that is worse than today's 10s. Caching the optimistic result bounds the cost at one penalty per tool per server lifetime. A non-timeout `execution_error` — spawn failure, permission denied — is not a timeout and still caches `False`.

- **Predicate order is load-bearing.** A timeout sets `execution_error` as well as `timed_out`, so today's `return_code == 0 and not result.execution_error` already excludes timeouts. Adding the fail-open branch *after* it leaves that branch unreachable: the change compiles, the existing tests pass, and the bug survives untouched. Branch on `timed_out` first.

- **`PYTEST_DISABLE_PLUGIN_AUTOLOAD=1` is safe on the probe** because it never touches the real run, so `-n auto` and xdist are unaffected. Set it uniformly on all probes: it is inert for the other four, and a per-tool exception costs more code than it saves. `execute_command(env=...)` is additive — `prepare_env` merges over an `os.environ` copy — so `PATH` is not clobbered. Its value is no longer speed but *finishing*: 5.2s against a 30s budget yields a definite answer where 15.2s on a slower machine yields a fail-open guess.

- **Third consumer of `venv_path`, easy to miss:** pytest's `PATH` prepend at `code_checker_pytest/runners.py:202-206`. The value becomes `Path(_resolved_python).parent`, but the parameter's *meaning* changes from venv root to bin directory — the function currently appends `Scripts`/`bin` itself, so the resolved script directory cannot simply be passed in as `venv_path`. It touches two signatures (`runners.py:89` and `check_code_with_pytest` at `runners.py:477`) and both docstrings, and `check_code_with_pytest` passes it to `run_tests` **positionally** (`runners.py:516-528`), so the rename lands in **three** places, not two — plus the structured log field at `runners.py:132` (`"venv_path": venv_path`) and the explanatory comment at `:151`. `tests/test_code_checker/test_runners.py:229` also passes it positionally. Passing `parent.parent` instead is wrong for a non-venv interpreter (`C:\Python311\python.exe` → `C:\`).

- **`--venv-path` keeps resolving the interpreter.** Only detection and the `PATH` prepend stop using it. The epilog example at `main.py:30` passes `--venv-path` without `--python-executable`; ignoring it for resolution too would silently point that configuration at `sys.executable` — trading a loud false negative for a quiet one.

- **Leaving the messages alone relocates the misdiagnosis.** Sixteen strings across eleven files name `--venv-path`: `server.py:138,157,174,191,208,232`, `ruff_check_tool.py:45`, `ruff_fix_tool.py:46`, `vulture_tool.py:44`, `bandit_tool.py:46`, `tach_tool.py:33`, `lint_imports_tool.py:43`, `mypy_tool.py:64`, `pylint_tool.py:42`, `pytest_tool.py:73`, `formatter_tools.py:69`. After the change every one points at a flag that no longer affects the outcome. The five `server.py` warnings are absorbed by the loop; the rest are separate edits.

- **The six script-tool messages print `"N/A"`.** `server._ruff_binary or "N/A"` — when the tool is unavailable that path is always `None`, so the message reads "ruff is not available at N/A". Report the directory searched (`Path(_resolved_python).parent`) instead; that is what would have made the original bug self-diagnosing.

- **Validation moves and improves.** `--venv-path` currently raises `FileNotFoundError` at startup if its python is missing (`server.py:105`); `--python-executable` gets no existence check at all. One check on the resolved interpreter replaces it, naming the flag that supplied the path. This is a new hard failure for a stale `--python-executable`, deliberately: a server whose interpreter does not exist cannot run a single tool, so starting up only defers the same conclusion across fifteen tool messages. `sys.executable` always exists, so the default configuration is unaffected.

- **Fail-open degrades unevenly; pytest is the weak spot.** mypy (`code_checker_mypy/runners.py:145`) and pylint (`code_checker_pylint/runners.py:106`) already call `check_tool_missing_error`, which turns `No module named X` into a clean message. black/isort surface a raw traceback. pytest fails via JSON-report parsing and reports worst. Worth checking during implementation; not a blocker.

- **Timeout stays a constant, deliberately.** No new CLI flag — #219 owns timeout configurability, and after this change the probe is off the common path entirely.

- **History:** #155 made these checks eager and parallel; #167 deferred them to first use. The lazy design is intentional — do not revert it by moving everything back to startup. The file-existence group stays eager; it is instant.

- **Three tests must be rewritten, not re-pointed.** `test_all_tools_missing`, `test_lint_imports_unavailable_when_no_venv` and `test_vulture_unavailable_when_no_venv` build a server with no `venv_path` and no `os.path.exists` patch, then assert all five script tools are unavailable. Once Decision 1 derives detection from `_resolved_python` (= `sys.executable` = the project's own venv under test), they will find the real binaries and flip to `True`. The replacements must pin the searched directory rather than depending on the ambient interpreter.

- **Decision 12 breaks two more tests.** `test_python_executable_fallback` passes `/usr/local/bin/python3.11` and `test_resolved_python_passed_to_pytest_runner` passes `/custom/python`; neither patches `os.path.exists`, so both raise `FileNotFoundError` once the resolved interpreter is validated.

- **A third strict assertion breaks on Decision 10.** `tests/test_server_params.py:74-85` calls `assert_called_once_with(..., venv_path=None, ...)` on `check_code_with_pytest`; the keyword rename lands there as well as in `tests/test_code_checker/test_runners.py:221-233`.

- **The mcp-config follow-up includes a test, not just a `ParameterDef`.** `tests/test_config/test_filesystem_venv_fix.py::test_code_checker_keeps_venv_path` asserts `"--venv-path" in config["args"]` for `MCP_CODE_CHECKER`. Still out of scope here; noted so it is not missed later.

- **Tests:** `tests/test_tool_availability.py::TestIsToolAvailable` is the existing home, and `tests/conftest.py::make_command_result` already accepts `timed_out`. At least seven existing tests need work: three rewritten (the ambient-interpreter group above), two fixed for Decision 12, and two re-pointed — `test_venv_path_not_found_raises` (now the resolved-interpreter check) and `tests/test_code_checker/test_runners.py:229` (positional `venv_path`).

## Decisions

| # | Topic | Decision |
|---|---|---|
| 1 | Detection source | Derive the script directory from `_resolved_python` everywhere; stop keying on `venv_path`. Applies to both the probe group and the eager group. |
| 2 | `--venv-path` | Soft-deprecate: still accepted, ignored for detection, logs a warning, hidden from `--help`. No hard removal — both `.mcp.json` files pass it and mcp-config still generates it. |
| 3 | Fallback order | File-existence fast path → `python -m` probe → fail open on timeout. Not pure file existence. |
| 4 | Timeout caching | Cache as *available* and log a WARNING. `_tool_availability` stays `dict[str, bool]`; no tri-state. |
| 5 | Probe environment | `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1` on all probes, uniformly. |
| 6 | Timeout value | Plain `30` constant, no new flag. |
| 7 | Descriptor `module` | `None` for lint-imports, vulture, ruff, bandit, tach — file existence is their entire check, no probe and no fail-open. Only pytest/pylint/mypy/black/isort probe. Supersedes any reading of #2 above that gives lint-imports `module=importlinter`. |
| 8 | Binary storage | `self._tool_binaries: dict[str, str]` replaces the five `_X_binary` attributes, and holds the **five script tools only** — pytest/pylint/mypy/black/isort are always invoked as `python -m <module>` and never via the console script, so entries for them would be written and never read. A loop cannot assign five distinct attribute names without `setattr`, which mypy strict and vulture both flag. Blast radius: **12 reads across 6 source files** (each tool module reads its `_X_binary` twice — once for the unavailable message, once for the run), plus **16 references across 3 test files** — `tests/test_checker_tools.py:21,34,35,36,37` (the `mock_server` fixture sets all five), `tests/test_tool_availability.py:162,175,178,203,205,207,222,234,236,542`, and `tests/test_code_checker_bandit/test_integration.py:16`. |
| 9 | Messages | All sixteen `--venv-path` mentions rewritten to name `--python-executable`; the six script-tool messages report the directory searched instead of `"N/A"`. |
| 10 | pytest `PATH` | Runner parameter becomes a bin directory (`venv_bin`); the `Scripts`/`bin` join moves to the call site — `src/mcp_tools_py/checker_tools/pytest_tool.py:106` (`venv_path=server.venv_path`) — as `Path(_resolved_python).parent`. Not `parent.parent`, and the prepend is not removed. Note for the implementer: `runners.py:204/206` already binds a local named `venv_bin`, the name given here to the new parameter; the parameter subsumes that local rather than conflicting with it, but the collision will be met during the edit. |
| 11 | `--venv-path` scope | Still resolves the interpreter with today's precedence. Only its detection and `PATH` roles are dropped, so `--venv-path`-only configurations keep working. |
| 12 | Interpreter validation | One startup `FileNotFoundError` on whatever `_resolve_python_executable` returns, naming the flag that supplied it. Replaces the venv-only check. Both `.mcp.json` files build the `--python-executable` value by interpolating an environment variable — `${VIRTUAL_ENV}` in this repo, `${MCP_CODER_VENV_DIR}` in mcp-config — so a client launching the server with that variable unset interpolates a malformed path, and that configuration now fails at startup instead of degrading to fifteen tool-level messages. Accepted: the startup error names the flag that supplied the path, which is more diagnosable than fifteen downstream messages, and making the failure mode depend on *which* flag supplied the path would reintroduce exactly the flag-keyed coupling this issue removes. `sys.executable` always exists, so the default configuration is unaffected. |
| 13 | Fail-open predicate | `timed_out` only, checked first. A spawn failure or other `execution_error` still caches `False`. |
| 14 | Documentation | In scope. `README.md` (parameter table `:105`, precedence bullet `:122`, config examples `:129,:142,:160`, troubleshooting `:171` and `:172`), `docs/architecture/architecture.md:213` and the `main.py:30` epilog example all lead with `--python-executable`. The README troubleshooting entry currently recommends `--venv-path` as the fix for "No module named pytest" — the same misdiagnosis this issue exists to remove, in a different file. One line below it, `README.md:172` claims "Tool availability is checked at startup and cached for the session" — untrue since #167 made the probe group lazy, and stale in exactly the way this decision targets. The epilog is printed as part of `--help`, so hiding the flag with `SUPPRESS` while the epilog advertises `--venv-path .venv` as a headline example would be self-contradictory. |
| 15 | Detection filesystem API | `os.path.exists`, not `pathlib`. Eight existing tests, all in `tests/test_tool_availability.py`, patch `mcp_tools_py.server.os.path.exists` (14 patch statements in total, since six of those same tests also patch `mcp_tools_py.server.os.name`); switching the new detection loop to `Path(...).exists()` would silently defeat every one of those patches, leaving the tests green while exercising the real filesystem. |

## Dependencies / references

- **Blocked on #219** — it must merge first. #219 rewrites subprocess timeout handling across the same files: `server.py` (adds `resolve_timeout` immediately after `_is_tool_available`), `main.py` (adds `--check-timeout`), all eight `checker_tools/*_tool.py`, `tests/test_checker_tools.py`, `tests/test_server_params.py`, `README.md` and `docs/architecture/architecture.md`. The sharpest conflict is one argument list in `tests/test_code_checker/test_runners.py` — #219 edits line 231, #220 edits line 229. The conflicts are mechanical, not semantic: #219 explicitly leaves the availability probe hardcoded, so Decision 6 (plain `30`, no flag) stays consistent. #220 is sequenced after #219 merges; rebasing #220 onto a merged #219 is far cheaper than the reverse. #220's `_is_tool_available` rewrite must preserve `resolve_timeout`, which will already exist by then.

- **mcp-config follow-up** (MarcusJellinghaus/mcp-config#56) — still emits `--venv-path` (`src/mcp_config/servers.py:515-522`) and asserts it in a test (see the constraint above). Out of scope here; the deprecation warning fires on every start until that repo is updated.

**Comment by MarcusJellinghaus** [comment]
Branch could not be rebased cleanly — handing off.
